torch/distributions/transforms.py

```diff
@@ -0,0 +1,242 @@
+import torch
+from torch.distributions import constraints
+from torch.distributions.utils import broadcast_all
+from torch.nn.functional import sigmoid
+
+__all__ = [
+    'AbsTransform',
+    'AffineTransform',
+    'ExpTransform',
+    'InverseTransform',
+    'LogprobTransform',
+    'SigmoidTransform',
+    'StickBreakingTransform',
+    'Transform',
+]
+
+
+class Transform(object):
+    """
+    Abstract class for transformations with computable inverse log
+    det jacobians. They are primarily used in
+    :class:`torch.distributions.TransformedDistribution`.
+
+    Transforms are intended to be short-lived objects. They memoize the forward
+    and inverse computations to avoid work; therefore :meth:`inverse` is
+    nearly free after calling :meth:`forward`. To clear the memoization cache,
+    delete the object and create a new object.
+
+    Derived classes should implement one or both of :meth:`_forward` or
+    :meth:`_inverse` and should implement :meth:`log_abs_det_jacobian`.
+    Derived classes may store intermediate results in the `._cache` dict.
+    """
+
+    def __init__(self):
+        self._cache = {}
+
+    def __eq__(self, other):
+        return type(other) is type(self)

```

Comment:
That's probably a bad idea. Think what happens when someone implements a transform that is somehow parametrized, but forgets to implement `__eq__`

Reply:
We may do something like this: introduce a flag for non-parameterized transforms, default `__eq__` to check for types if `self` and `other` are both non-parameterized, otherwise raise `NotImplementedError`
 
For parameterized transforms, I have no idea how to handle batches.

Reply:
We don't need to handle batches; that sounds a bit over-engineered. In fact I think it's safe to define `Transform.__eq__` by `self is other` by default and define `type(self) == type(other)` for the each of few parameter-free transforms. WDYT?